### DISCUSSION TITLE
Compile-time constants to replace preprocessor macros in configuration files?

### DISCUSSION THREAD

**Discussion by ilg-ul** [carbon-language/carbon-lang#2052] [Language design]
I'm considering extending one of my projects with a configuration tool inspired by eCos CDL (to configure C++ programs), and, looking to the future, I'm trying to figure out how such a configuration tool is expected to work with Carbon.

In C/C++ the usual way to pass configuration data is via an `#include "config.h"` where several preprocessor macros are defined, and later the code can test for the macros presence or use the values to define the size of arrays, etc.

In Carbon, which, if I got it right, does not have a preprocessor, I expect such configurations to be implemented with compile-time constants, booleans for simple yes/no choices, integer for sizes, etc.

However, I was not able to figure out how this mechanism is expected to work... :-(

Could someone provide an example how a file with such configuration constants may look like, and how it should be included in other files?

Thank you in advance.

**Comment by jonmeow** [comment]
This isn't a design that's been fully designed out yet, but metaprogramming will exist in some form, and this kind of use case would probably [use let](https://github.com/carbon-language/carbon-lang/tree/trunk/docs/design#constant-let-declarations), e.g. `let HasFooApi: bool = false;`. Whether there's a separate `constexpr` equivalent keyword isn't something that I think has been discussed.

**Reply by ilg-ul** [reply]
Thank you, Jon. I tried to find more details about Carbon metaprogramming, but the documentation page has no content yet.

As for the actual definitions syntax, I guess it will not be a problem for an automated tool to generate such configuration files, regardless of the final syntax details.

One more question: how is it planned to include such configuration files in regular source files?

**Reply by L4stR1t3s** [reply]
> Whether there's a separate constexpr equivalent keyword isn't something that I think has been discussed.

What would you do if you don't have a separate constexpr keyword? Consider everything that is const as constexpr as well? That won't work IMO because not everything that is constant at run-time is also constant at compile-time. You could go pretty far with that, and make a lot of code work at compile-time, but not everything:
`const auto id = glCreateShader(GL_VERTEX_SHADER); // Create an OpenGL vertex shader, returns id of the shader on the GPU`
Even if you made glCreateShader work at compile-time (no technical reason you couldn't) it would be pretty useless since the id would not be valid at run-time.
And how would you determine whether to evaluate something at compile-time, or at run-time? I would like to be able to generate random numbers at compile-time for example, but definitely not all random numbers. If the same random number function can be used at compile-time and run-time, then some sort of distinction would have to be made syntax-wise.
Or this would mean that the same function could not be used but we need separate functions, but it would be rather cumbersome to have to define all functions twice if you want to use them at compile-time and run-time.

**Reply by ilg-ul** [reply]
Even if the details are not yet clear, compile-time constants are mentioned in the design documents, so I expect them to be available, in one form or another, otherwise metaprogramming will not be complete.

**Reply by jonmeow** [reply]
@ilg-ul Similar to having it be a `.h` file for C++, I'd assume the starting point would be `.carbon` file for Carbon and using standard inclusion mechanisms. I don't think see any reason why this would be special-cased.

@L4stR1t3s While you raise some good points, I don't really have any references to point you to because metaprogramming and compile-time evaluation just hasn't been a priority for the design work done so far.

**Reply by L4stR1t3s** [reply]
> @L4stR1t3s While you raise some good points, I don't really have any references to point you to because metaprogramming and compile-time evaluation just hasn't been a priority for the design work done so far.

@jonmeow That makes me a little uneasy, because meta programming and compile time evaluation are going to become increasingly more important in the future, and for Carbon to succeed it will have to make a real difference in those areas IMO. The issues with meta programming and compile-time evaluation in C++ are often related to these features being added to an already existing language. I hope Carbon doesn't run into similar issues in the future by assigning a low priority now. Or at least that in the future decisions made now can and will be overturned if it turns out meta programming and compile-time evaluation are negatively impacted by them.

**Reply by ilg-ul** [reply]
> Similar to having it be a .h file for C++, I'd assume the starting point would be .carbon file for Carbon and using standard inclusion mechanisms. I don't think see any reason why this would be special-cased.

@jonmeow, I tried to find some more details in the proposals, and in p0107 there are long arguments related to importing packages and libraries, but apparently there is nothing similar to passing `.h` files to C/C++. 

I don't say that using packages/libraries for passing configuration constants is wrong or impossible, just that there is no discussion explaining how it is supposed to be.

Also in p0339 it is said _this proposal does not include any syntax for constants. This is expected to be revisited later._

There might be other interesting things in the proposals, but since they are quite difficult to browse through, I did not find them.

So, it looks like the idea to fully replace the preprocessor with metaprogramming is only a desire, not a concrete proposal. :-(

**Comment by L4stR1t3s** [comment]
I think there is a speed and memory issue with using compile-time constants as a replacement for preprocessor macros. Preprocessor macros are evaluated before compilation, meaning that potentially a lot of code is removed before compilation. If the decision of which code to use or not is deferred to the compilation then this will have a negative impact on speed and memory while compiling.

**Reply by ilg-ul** [reply]
Right, but apparently the decision to not include a preprocessor has been already made, so I'm trying to figure out how things are expected to work without it.

Personally I completely avoid macros in C++, but I think that configuration constants and conditional compilations are features that should be present in any language that plans to compete in the space of C++ applications, especially embedded applications, which require an even higher degree of compile-time configurability.

**Reply by mossaiby** [reply]
Of course C/C++ pre-processor does much more than just `#ifdef` and `#ifndef`, so compile-time constants cannot replace current functionality on their own. Configuration problem is so simple compared to advanced use of pre-processor. What can be done for those cases?

**Reply by L4stR1t3s** [reply]
> Personally I completely avoid macros in C++

I have a hard time believing that. Unless our definition of a macro is not the same? In C++ anything that uses #define is considered to be a macro as it will be replaced whenever it is found by the preprocessor.
`#define FLAG` I am a macro!
`#define VERSION 1` I am a macro! 
`#define SUM(a, b) a + b` I am ~~Spartacus~~ a macro!
I can believe someone not using the third macro style, but never using the first and second macro style? I would be flabbergasted!

I understand the sentiment, macros in C++ are one of the worst features of the language IMO, but sadly I don't think they can entirely be avoided when using C++.

> What can be done for those cases?

I also fear that Carbon can not entirely avoid a pre-processing step, as it is the most efficient way of dealing with code configuration IMO. But I am aware that opening that door is dangerous, because who knows what might sneak in as well.

**Reply by ilg-ul** [reply]
> I have a hard time believing that. 

Technically you are right, all preprocessor definitions are called _macros_, but I differentiate between _object-like macros_, which I accept for configuration constants, and _function-like macros_, which I avoid at all cost. 

> I also fear that Carbon can not entirely avoid a pre-processing step, as it is the most efficient way of dealing with code configuration IMO

I am a bit confused now, initially I thought the same, but when I read in the proposals that the preprocessor functionality will be covered by metaprogramming, I thought that perhaps the design team had some inovative ideas how to do it. 

Now I'm not so sure, it looks like the subject was not even considered thoroughly. :-(

**Reply by mossaiby** [reply]
> Technically you are right, all preprocessor definitions are called macros, but I differentiate between object-like macros, which I accept for configuration constants, and function-like macros, which I avoid at all cost.

Not all the things can be done without _object-like macros_, and many things than can be done, need weird C++ tricks that make the code much harder to read and understand (at least for me). Yet, I agree pre-processor is a two-way sword, but I do not have a definite replacement for it.
